New wavefilewriter breaks build on s390x, ppc64 and sparc64

Hi!

The recently introduced wavefilewriter breaks the build on the s390x, ppc64 and sparc64 architectures in Debian [1, 2, 3]:

g++ -m64 -specs=/usr/share/dpkg/pie-link.specs -Wl,-z,relro -specs=/usr/share/dpkg/pie-link.specs -Wl,-z,relro -o qhimdtransfer main.o qhimdaboutdialog.o qhimdformatdialog.o qhimduploaddialog.o qhimdmainwindow.o qhimddetection.o qmdmodel.o qmdtrack.o qmddevice.o qhimddummydetection.o wavefilewriter.o moc_qhimdaboutdialog.o moc_qhimdformatdialog.o moc_qhimduploaddialog.o moc_qhimdmainwindow.o moc_qhimddetection.o moc_qmdmodel.o moc_qmddevice.o qrc_icons.o    -L../libhimd -L../libnetmd -L/usr/lib/sparc64-linux-gnu -L../libhimd -lhimd -lmad -L../libnetmd -lnetmd -lgcrypt -lglib-2.0 -lusb-1.0 -ltag -lQtGui -lQtCore -lpthread 
wavefilewriter.o: In function `unsigned long qToLittleEndian<unsigned long>(unsigned long)':
/usr/include/qt4/QtCore/qendian.h:342: undefined reference to `unsigned long qbswap<unsigned long>(unsigned long)'
collect2: error: ld returned 1 exit status

It seems a very similar issue was fixed in goldendict [4].

This is something we have to fix as otherwise linux-minidisc will not be shipped in Debian Stretch!

Adrian

> [1] https://buildd.debian.org/status/fetch.php?pkg=linux-minidisc&arch=s390x&ver=0.9.14-1&stamp=1481283541
> [2] https://buildd.debian.org/status/fetch.php?pkg=linux-minidisc&arch=ppc64&ver=0.9.14-1&stamp=1481285857
> [3] https://buildd.debian.org/status/fetch.php?pkg=linux-minidisc&arch=sparc64&ver=0.9.14-1&stamp=1481283862
> [4] https://github.com/LocutusOfBorg/goldendict/commit/c035954693461c135f9685253f2b736946376634

--- a/qhimdtransfer/wavefilewriter.cpp
+++ b/qhimdtransfer/wavefilewriter.cpp
@@ -126,7 +126,7 @@ updateU32LESizeField(QFile &file, size_t offset, uint32_t new_value)
     }
 
     // The "remaining" size is calculated from after the value (subtract offset + value size)
-    new_value = qToLittleEndian(new_value - (offset + sizeof(new_value)));
+    new_value = qToLittleEndian(uint32_t(new_value - (offset + sizeof(new_value))));
     if ((size_t)file.write((char *)&new_value, sizeof(new_value)) != sizeof(new_value)) {
         qWarning() << "Could not update field in file";
         return false;

